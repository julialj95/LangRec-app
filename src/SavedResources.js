import React, { Component } from "react";
import { LangrecContext } from "./LangrecContext";
import ResultItem from "./ResultItem";
import TokenService from "./services/token-service";
import config from "./config";

class SavedResources extends Component {
  static contextType = LangrecContext;
  constructor() {
    super();
    this.state = {
      results: [],
    };
  }

  displayResults = (data) => {
    console.log(data);
    const results = data.map((result) => {
      return (
        <div>
          <ResultItem
            key={result.id}
            id={result.id}
            title={result.title}
            image_link={result.image_link}
            language={result.language}
            level={result.level}
            type={result.type}
            rating={result.rating}
            url={result.url}
            description={result.description}
            cost={result.cost}
          />
          <button
            value={result.id}
            key={result.url}
            onClick={this.removeFromFavorites}
          >
            Remove resource from favorites
          </button>
        </div>
      );
    });
    this.setState({ results });
    return results;
  };

  removeFromFavorites = (e) => {
    e.preventDefault();
    const resource_id = e.target.value;
    fetch(`${config.API_BASE_URL}/resources/saved-resources/${resource_id}`, {
      method: "DELETE",
      headers: {
        "content-type": "application/json",
        Authorization: "Bearer " + TokenService.getAuthToken(),
      },
    })
      .then((res) => {
        if (!res.ok) return res.json().then((e) => Promise.reject(e));
        const newResults = this.state.results.filter(
          (resource) => resource.id !== resource_id
        );
        this.setState({ results: newResults });
      })
      .catch((error) => console.error(error));
  };

  componentDidMount() {
    fetch(`${config.API_BASE_URL}/resources/saved-resources`, {
      headers: {
        "content-type": "application/json",
        Authorization: "Bearer " + TokenService.getAuthToken(),
      },
    })
      .then((response) => {
        if (!response.ok) {
          return response.json().then((event) => Promise.reject(event));
        }
        return response.json();
      })
      .then((data) => {
        this.displayResults(data);
      })
      .catch((error) => console.error({ error }));
  }

  render() {
    return (
      <div>
        <div>
          {this.state.results.length === 0 ? (
            <div>
              <h3>You have no saved resources.</h3>
              <h3>
                Search through resource recommendations and find resources to
                save!
              </h3>
            </div>
          ) : (
            this.state.results
          )}
        </div>
      </div>
    );
  }
}

export default SavedResources;
