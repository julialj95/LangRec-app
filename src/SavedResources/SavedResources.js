import React, { Component } from "react";
import { LangrecContext } from "../LangrecContext";
import ResultItem from "../ResultItem/ResultItem";
import TokenService from "../services/token-service";
import config from "../config";

class SavedResources extends Component {
  static contextType = LangrecContext;
  constructor() {
    super();
    this.state = {
      results: [],
    };
  }

  displayResults = () => {
    console.log(this.context.savedResources);
    const results = this.context.savedResources.map((result) => {
      return (
        <div key={result.image_link}>
          <ResultItem
            key={result.id}
            id={result.id}
            title={result.title}
            image_link={result.image_link}
            language={result.language}
            level={result.level}
            type={result.type}
            rating={result.rating}
            url={result.url}
            description={result.description}
            cost={result.cost}
          />
          <button
            key={result.title}
            value={result.id}
            onClick={this.deleteFromFavorites}
          >
            Remove resource from favorites
          </button>
        </div>
      );
    });

    if (this.context.savedResources.length === 0) {
      return (
        <div>
          <h3>You have no saved resources.</h3>
          <h3>
            Search through resource recommendations and find resources to save!
          </h3>
        </div>
      );
    } else {
      return results;
    }
  };

  deleteFromFavorites = (e) => {
    e.preventDefault();
    const resource_id = e.target.value;
    fetch(`${config.API_BASE_URL}/resources/saved-resources/${resource_id}`, {
      method: "DELETE",
      headers: {
        "content-type": "application/json",
        Authorization: "Bearer " + TokenService.getAuthToken(),
      },
    })
      .then((res) => {
        if (!res.ok) return res.json().then((e) => Promise.reject(e));
      })
      .then(() => {
        this.context.removeSavedResource(resource_id);
      })
      .catch((error) => console.error(error));
  };

  componentDidMount() {
    fetch(`${config.API_BASE_URL}/resources/saved-resources`, {
      headers: {
        "content-type": "application/json",
        Authorization: "Bearer " + TokenService.getAuthToken(),
      },
    })
      .then((response) => {
        if (!response.ok) {
          return response.json().then((event) => Promise.reject(event));
        }
        return response.json();
      })
      .then((data) => {
        this.context.getSavedResources(data);
      })
      .catch((error) => console.error({ error }));
  }

  render() {
    return <div>{this.displayResults()}</div>;
  }
}

export default SavedResources;
