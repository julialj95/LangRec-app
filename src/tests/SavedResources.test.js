import React from "react";
import ReactDOM from "react-dom";
import SavedResources from "../SavedResources/SavedResources";
import { BrowserRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";

it("renders without crashing", () => {
  const div = document.createElement("div");

  ReactDOM.render(
    <Router>
      <SavedResources />
    </Router>,
    div
  );

  ReactDOM.unmountComponentAtNode(div);
});

it("renders the UI as expected", () => {
  const tree = renderer
    .create(
      <Router>
        <SavedResources />
      </Router>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
