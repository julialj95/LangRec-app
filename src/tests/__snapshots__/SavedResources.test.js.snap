// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the UI as expected 1`] = `
Array [
  <h2>
    SAVED RESOURCES
  </h2>,
  <div>
    <div>
      <h3>
        You have no saved resources.
      </h3>
      <h3>
        Search through 
        <a
          href="/recs"
          onClick={[Function]}
        >
          resource recommendations
        </a>
         and find resources to save!
      </h3>
    </div>
  </div>,
]
`;
